### DISCUSSION TITLE
Questions about Select with Modal and Polymorphism

### DISCUSSION THREAD

**Discussion by antoinemy** [filamentphp/filament#5368] [Help] [answered]
Hi there,

I have a few questions about forms and particularly Select, I couldn't find anything in the documentation.

Here is an example :

```
Select::make('my_select')
    ->relationship('my_relation', 'updated_at')
    ->getOptionLabelFromRecordUsing(fn (Model $record) => "{$record->name} - {$record->updated_at->diffForHumans()}")
    ->createOptionForm([
      // Inputs in my modal
    ])
    ->createOptionAction(function() {
       // How can use this? Any documentation?
    })
    ->createOptionUsing(function() {
        // How can use this? Any documentation?
    }),
```
**Question 1 :** Can I change the size of the Creation Modal? if yes, how?

**Question 2 :** If I want to change or remove a selected item from my Select, how can I implement it?

Then I have a Repeater, which has a Select and this Select has elements who are polymorphs:

```
Repeater::make('my_repeater')
    ->schema([

        Select::make('my_select')
            ->options(ProductSupplied::where('type_id', 10)->get()->pluck('product.name', 'id'))
            //->getOptionLabelFromRecordUsing(fn (Model $record) => "{$record->product->name} - {$record->updated_at->diffForHumans()}")
            ->createOptionAction(function() {
               // How can use this? Any documentation?
            })
            ->createOptionUsing(function() {
                // How can use this? Any documentation?
            })
            ->createOptionForm([
                // Inputs in modals
            ]),
      ])
```

**Question 3:** The function getOptionLabelFromRecordUsing() (which is commented out) doesn't work since I don't have a function Relationship(). How can I customize my Select information without this function and with polymorphism?

**Question 4:** Since I don't have a function Relationship() on my Select since it's in a Repeater, the modal's form doesn't save anything and I don't know where I can take the control? So I have a problem with the form submission and with the polymorph. And I think the two are related.

______

To make it simpler, I want to have a Select and to its right add, modify and delete buttons. While opening a Modal to add an item in the Select, modify the selected item in the list in a Modal and delete with a confirmation request by Modal.

To add an element in the list, your option of adding by Modal works very well. And that suits me! But I lack documentation to control the UI of the Modal and to use it with polymorphism.

To edit and delete in the list, I imagine I'm going to have to create a custom component?

If you have any solutions,
Many thanks for the work you have done,
I can't wait to see V3!

**Comment by danharrin** [comment, marked as answer]
1) `createOptionAction(fn ($action) => $action->modalWidth('4xl'))` etc. `createOptionAction()` is for modifying the action object.

You don't need `createOptionUsing()` if youre using `relationship()`, we set it up for you.

2) I don't know what I would suggest for editing / deleting options. I can't imagine a UI where this would be very convenient. Why not just maintain a separate interface (a table or something) to handle this?

3) Instead of using `pluck()`, chose your own way to transform the models into labels. `getOptionLabelFromRecordUsing()` is used when we get the labels for you. But you're passing them youself. Try `mapWithKeys()` or something on the collection.

4) That's where `createOptionUsing()` comes in, you can access the form `$data` and return the key of the new select option.

**Comment by antoinemy** [comment]
How can I do to select the new item in the Select after it is created with createOptionUsing() like an autofocus?

```
->createOptionForm([
  Grid::make(['default' => 12])->schema([
      TextInput::make('name')
          ->columnSpan([
              'default' => 12,
              'md' => 12,
              'lg' => 12,
          ]),
  ]),
])
->createOptionUsing(function($data) {
  $data = (object) $data;
  Product::create([
      'product_category_id' => 4,
      'name' => $data->name,
  ]);
  Notification::make()
      ->title('Saved successfully')
      ->icon('heroicon-o-document-text') 
      ->iconColor('success') 
      ->send();
})
```

**Reply by danharrin** [reply]
Return the product ID after its created

**Reply by antoinemy** [reply]
Thank's, it works good! :)
